File: exam_app/lib/models/exam_event.dart
import 'package:json_annotation/json_annotation.dart';

part 'exam_event.g.dart';

@JsonSerializable()
class ExamEvent {
  final ExamEventType type;
  final int? examId;
  final int? userId;
  final dynamic data;
  final DateTime timestamp;

  ExamEvent({
    required this.type,
    this.examId,
    this.userId,
    this.data,
    required this.timestamp,
  });

  factory ExamEvent.fromJson(Map<String, dynamic> json) => _$ExamEventFromJson(json);
  Map<String, dynamic> toJson() => _$ExamEventToJson(this);
}

enum ExamEventType {
  @JsonValue('USER_JOINED')
  userJoined,
  @JsonValue('USER_LEFT')
  userLeft,
  @JsonValue('ANSWER_SUBMITTED')
  answerSubmitted,
  @JsonValue('EXAM_COMPLETED')
  examCompleted,
  @JsonValue('TIME_WARNING')
  timeWarning,
  @JsonValue('EXAM_ENDED')
  examEnded,
  @JsonValue('STATISTICS_UPDATED')
  statisticsUpdated,
}

File: exam-api/src/main/java/com/camoleze/examapi/dto/ExamEvent.java
package com.camoleze.examapi.dto;

import com.fasterxml.jackson.annotation.JsonFormat;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class ExamEvent {
    private ExamEventType type;
    private Long examId;
    private Long userId;
    private Object data;

    @JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss")
    private LocalDateTime timestamp;
    
    public enum ExamEventType {
        USER_JOINED,
        USER_LEFT,
        ANSWER_SUBMITTED,
        EXAM_COMPLETED,
        TIME_WARNING,
        EXAM_ENDED,
        STATISTICS_UPDATED
    }
}
